# **Final Paper** (Python Language)

# **Mohsan Ali**


Name:         -----------------------           **Mohsan Ali**

Roll No.:     -----------------------           **122.20**

Class:        -----------------------           7th Sem.(Replica)

Section:        -----------------------         Bs Statistics

Subject:      -----------------------           Python Programming




# **Paper Start**


### Here I import all necessary Libraries of Python

In [148]:
# Import required libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import statsmodels.api as sm
from scipy.stats import f_oneway
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline



### Here I Import the BikeStoresales_data.csv dataset using numpy library

In [149]:
# Here I Import the BikeStoresales_data.csv dataset using numpy library pandas And Load the dataset into a pandas DataFrame

dataset = pd.read_csv("C:/Users/Joyia/OneDrive/Desktop/paper/BikeStoresales_data.csv")
dataset


Date  Day     Month  Year  Customer_Age       Age_Group  \
0       11/26/2013   26  November  2013            19     Youth (<25)   
1       11/26/2015   26  November  2015            19     Youth (<25)   
2        3/23/2014   23     March  2014            49  Adults (35-64)   
3        3/23/2016   23     March  2016            49  Adults (35-64)   
4        5/15/2014   15       May  2014            47  Adults (35-64)   
...            ...  ...       ...   ...           ...             ...   
113031   4/12/2016   12     April  2016            41  Adults (35-64)   
113032    4/2/2014    2     April  2014            18     Youth (<25)   
113033    4/2/2016    2     April  2016            18     Youth (<25)   
113034    3/4/2014    4     March  2014            37  Adults (35-64)   
113035    3/4/2016    4     March  2016            37  Adults (35-64)   

       Customer_Gender         Country             State Product_Category  \
0                    M          Canada  British Columbia      Accessories   
1                    M          Canada  British Columbia      Accessories   
2                    M       Australia   New South Wales      Accessories   
3                    M       Australia   New South Wales      Accessories   
4                    F       Australia   New South Wales      Accessories   
...                ...             ...               ...              ...   
113031               M  United Kingdom           England         Clothing   
113032               M       Australia        Queensland         Clothing   
113033               M       Australia        Queensland         Clothing   
113034               F          France     Seine (Paris)         Clothing   
113035               F          France     Seine (Paris)         Clothing   

       Sub_Category              Product  Order_Quantity  Unit_Cost  \
0        Bike Racks  Hitch Rack - 4-Bike               8         45   
1        Bike Racks  Hitch Rack - 4-Bike               8         45   
2        Bike Racks  Hitch Rack - 4-Bike              23         45   
3        Bike Racks  Hitch Rack - 4-Bike              20         45   
4        Bike Racks  Hitch Rack - 4-Bike               4         45   
...             ...                  ...             ...        ...   
113031        Vests      Classic Vest, S               3         24   
113032        Vests      Classic Vest, M              22         24   
113033        Vests      Classic Vest, M              22         24   
113034        Vests      Classic Vest, L              24         24   
113035        Vests      Classic Vest, L              23         24   

        Unit_Price  Profit  Cost  Revenue  
0              120     590   360      950  
1              120     590   360      950  
2              120    1366  1035     2401  
3              120    1188   900     2088  
4              120     238   180      418  
...            ...     ...   ...      ...  
113031          64     112    72      184  
113032          64     655   528     1183  
113033          64     655   528     1183  
113034          64     684   576     1260  
113035          64     655   552     1207  

[113036 rows x 18 columns]

### Question 01:   How many sales per month in the year (2013 & 2014) do we have?

In [150]:


# her I make a new column by name Date and Convert 'Date' column to datetime type
dataset['Date'] = pd.to_datetime(dataset['Date'])

# Filter sales data for the years 2013 and 2014
salesByMonth = dataset[(dataset['Year'] == 2013) | (dataset['Year'] == 2014)]

# Group by month and count the number of sales
sales_per_month = salesByMonth.groupby(['Year', 'Month']).size().reset_index(name='Sales Count')

print("Sales per month in the years 2013 & 2014:\n")
print(sales_per_month)


Sales per month in the years 2013 & 2014:

    Year      Month  Sales Count
0   2013      April          294
1   2013     August         3819
2   2013   December         5270
3   2013   February          272
4   2013    January          244
5   2013       July         1411
6   2013       June          321
7   2013      March          272
8   2013        May          335
9   2013   November         4174
10  2013    October         4146
11  2013  September         3885
12  2014      April         4590
13  2014   February         4068
14  2014    January         4210
15  2014       July         2067
16  2014       June         5082
17  2014      March         4366
18  2014        May         5015


### Question 02:   Which country has the most sales quantity of sales?

In [151]:

# Group by country and sum the sales quantity
sales_by_country = dataset.groupby('Country')['Order_Quantity'].sum().reset_index()

# Find the country with the maximum sales quantity
country_with_most_sales = sales_by_country.loc[sales_by_country['Order_Quantity'].idxmax()]

print("Country with the most sales quantity is:", country_with_most_sales['Country'])
print("With Total sales quantity:", country_with_most_sales['Order_Quantity'])



Country with the most sales quantity is: United States
With Total sales quantity: 477539


### Question 03: How many orders were made in each region (state) of France?

In [152]:
# Filter sales data for France
dataset_france = dataset[dataset['Country'] == 'France']
# Group sales data by state and count the number of orders
orders_per_region = dataset_france.groupby('State').size()

print("Number of orders made in each region (state) of France:\n")
print(orders_per_region)


Number of orders made in each region (state) of France:

State
Charente-Maritime     148
Essonne               994
Garonne (Haute)       208
Hauts de Seine       1084
Loir et Cher          120
Loiret                382
Moselle               386
Nord                 1670
Pas de Calais          90
Seine (Paris)        2328
Seine Saint Denis    1684
Seine et Marne        394
Somme                 134
Val d'Oise            264
Val de Marne          158
Yveline               954
dtype: int64


### Question 04. Get the top-5 sales with the highest revenue

In [153]:


# Sort the dataset by revenue in descending order
dataset_sorted = dataset.sort_values(by='Revenue', ascending=False)

# Select the top 5 sales with the highest revenue
top_5_sales = dataset_sorted.head(5)

print("Top-5 sales with the highest revenue:")
print(top_5_sales.T)


Top-5 sales with the highest revenue:
                                   112073                   112072  \
Date                  2015-07-24 00:00:00      2013-07-24 00:00:00   
Day                                    24                       24   
Month                                July                     July   
Year                                 2015                     2013   
Customer_Age                           52                       52   
Age_Group                  Adults (35-64)           Adults (35-64)   
Customer_Gender                         M                        M   
Country                         Australia                Australia   
State                          Queensland               Queensland   
Product_Category                 Clothing                 Clothing   
Sub_Category                        Vests                    Vests   
Product           Touring-1000 Yellow, 50  Touring-1000 Yellow, 50   
Order_Quantity                         29           

### Question 05. Are all the age groups providing equal average profit? apply the appropriate inferential analysis and interpret the results.

In [154]:
# Perform ANOVA to test if there is a significant difference in profit among age groups
# Assuming 'Age_Group' is the column containing age group information and 'Profit' is the column containing profit
groups = []
for age_group, group_data in dataset.groupby('Age_Group'):
    groups.append(group_data['Profit'])

# Perform one-way ANOVA
statistic, p_value = f_oneway(*groups)

# Set significance level
alpha = 0.05

print("Results of one-way ANOVA:")
print("F-statistic:", statistic)
print("p-value:", p_value)

if p_value < alpha:
    print("Reject the null hypothesis. There is significant evidence that at least one age group's mean profit is different.")
else:
    print("Fail to reject the null hypothesis. There is no significant evidence that the mean profit differs among age groups.\n")
    
print(f'Profit = -22185293586238.848 + 30261039228373.39 * {dataset.Order_Quantity} +  30261039228383.4 * {dataset.Age_Group} + 30261039228358.02 * {dataset.Product_Category} +  128265.85807291396')

Results of one-way ANOVA:
F-statistic: 66.88322142388175
p-value: 3.334502632863006e-43
Reject the null hypothesis. There is significant evidence that at least one age group's mean profit is different.
Profit = -22185293586238.848 + 30261039228373.39 * 0          8
1          8
2         23
3         20
4          4
          ..
113031     3
113032    22
113033    22
113034    24
113035    23
Name: Order_Quantity, Length: 113036, dtype: int64 +  30261039228383.4 * 0            Youth (<25)
1            Youth (<25)
2         Adults (35-64)
3         Adults (35-64)
4         Adults (35-64)
               ...      
113031    Adults (35-64)
113032       Youth (<25)
113033       Youth (<25)
113034    Adults (35-64)
113035    Adults (35-64)
Name: Age_Group, Length: 113036, dtype: object + 30261039228358.02 * 0         Accessories
1         Accessories
2         Accessories
3         Accessories
4         Accessories
             ...     
113031       Clothing
113032       Clothing
113033     

### Question 06. Use 80% of the data, to build a regression model for predicting the profit based on features such as order quantity, age group, and product category.

In [155]:
# Define features and target variable
features = ['Order_Quantity', 'Age_Group', 'Product_Category']
target = 'Profit'

# Split data into training and testing sets
train_data, test_data = train_test_split(dataset, test_size=0.2, random_state=42)

# Define preprocessing steps for categorical variables
preprocessor = ColumnTransformer(
    transformers=[
        ('cat', OneHotEncoder(), ['Age_Group', 'Product_Category'])
    ],
    remainder='passthrough'
)

# Define the regression model
model = Pipeline([
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression())
])

# Train the regression model
model.fit(train_data[features], train_data[target])

# Print intercept and coefficients
intercept = model.named_steps['regressor'].intercept_
coefficients = model.named_steps['regressor'].coef_

print("Intercept:", intercept)
print("Coefficients:")
for feature, coef in zip(features, coefficients):
    print(f"{feature}: {coef}")

# Evaluate the model
predictions = model.predict(test_data[features])
mse = mean_squared_error(test_data[target], predictions)
print("Mean Squared Error:", mse)


Intercept: -22185293586238.848
Coefficients:
Order_Quantity: 30261039228373.39
Age_Group: 30261039228383.4
Product_Category: 30261039228358.02
Mean Squared Error: 128265.85807291396


### Question 07. Use the rest of the data set to report the performance (R2, RMSE, etc.) of the regression model for predicting profit.

In [156]:


# Use the rest of the data set (the data not used for training) for evaluation
eval_data = dataset.drop(train_data.index)

# Predict profit using the regression model
predictions = model.predict(eval_data[features])

# Calculate R-squared
r_squared = r2_score(eval_data[target], predictions)

# Calculate RMSE
rmse = np.sqrt(mean_squared_error(eval_data[target], predictions))

print("Evaluation Metrics:")
print("R-squared:", r_squared)
print("Root Mean Squared Error (RMSE):", rmse)


Evaluation Metrics:
R-squared: 0.38857178696323036
Root Mean Squared Error (RMSE): 358.142231624412


# **The Overall Summery of our all analysis**

**Data Loading**: The code begins by loading the BikeStoresales_data dataset from a CSV file named 'BikeStoresales_data.csv' using the Pandas library.
**Libraries Loading**: then I import essential Libraries

**Average Sale**: Then I analyze the sales per month in the year (2013 & 2014)
**Most Sales by Country**: Then we find country has the most  quantity of sale which is:
United States With Total sales quantity: 477539

**Top Sales**: then we find the top sales of highest revenue

**Check equal average profit**: Then we Perform ANOVA to test if there is a significant difference in profit among age groups. which tell us that the **p value** is less then 0.5 so we reject our anova hypothesis.

## **Regression Model**

**Data Preprocessing**:
The dataset is split into features (independent variables) and the target variable, which is profit.

Categorical variables such as 'Age_Group' and 'Product_Category' are identified as features for the regression model.
where 80% of the data is used for training the regression model, and 20% is reserved for(testing) evaluating its performance.

**Model Building**:
 **Profit** = -22185293586238.848 + 30261039228373.39 * Order_Quantity +  30261039228383.4 * Age_Group + 30261039228358.02 * Product_Category +  128265.85807291396
The code preprocesses the categorical variables using one-hot encoding.
It then constructs a linear regression model using scikit-learn's Pipeline and LinearRegression classes.
The model is trained using the training data, including the order quantity, age group, and product category features to predict profit.

**Model Evaluation**:
The trained model is evaluated using the remaining 20% of the dataset that was not used for training.
Predictions are made on the evaluation dataset, and evaluation metrics such as R-squared (R2) and root mean squared error (RMSE) are calculated.
The R2 score quantifies the proportion of the variance in the target variable (profit) that is explained by the model, while RMSE provides a measure of the average prediction error.

**Output**:
The code prints out the evaluation metrics (R2 and RMSE) to assess the performance of the regression model on unseen data.


**Accuracy**:
The code performs data preprocessing, builds a regression model using 80% of the data, and evaluates the model's performance on the remaining 20% of the data. This approach allows for understanding how well the model generalizes to new, unseen data and provides insights into its predictive accuracy.